rust/src/storage/s3/mod.rs

```diff
@@ -362,13 +379,113 @@ impl StorageBackend for S3StorageBackend {
     }
 }
 
+/// A lock that has been successfully acquired
+#[derive(Clone, Debug)]
+pub struct LockItem {
+    /// The name of the owner that owns this lock.
+    pub owner_name: String,
+    /// Current version number of the lock in DynamoDB. This is what tells the lock client
+    /// when the lock is stale.
+    pub record_version_number: String,
+    /// The amount of time (in seconds) that the owner has this lock for.
+    pub lease_duration: u64,
+    /// Tells whether or not the lock was marked as released when loaded from DynamoDB.
+    pub is_released: bool,
+    /// Optional data associated with this lock.
+    pub data: Option<String>,
+    /// The last time this lock was updated or retrieved.
+    pub lookup_time: u128,
+}
+
+///  TODO
+#[async_trait::async_trait]
+pub trait LockClient: Send + Sync + Debug {
+    /// Attempts to acquire lock. If successful, returns the lock.
+    /// Otherwise returns [`Option::None`] which is retryable action.
+    /// Visit implementation docs for more details.
+    async fn try_acquire_lock(&self) -> Result<Option<LockItem>, StorageError>;
+
+    /// Returns current lock from DynamoDB (if any).
+    async fn get_lock(&self) -> Result<Option<LockItem>, StorageError>;
+
+    /// Releases the given lock if the current user still has it, returning true if the lock was
+    /// successfully released, and false if someone else already stole the lock
+    async fn release_lock(&self, lock: &LockItem) -> Result<bool, StorageError>;
+}
+
+#[allow(clippy::unnecessary_wraps)]
+fn try_create_lock_client(_region: Region) -> Result<Option<Box<dyn LockClient>>, StorageError> {
+    match std::env::var("AWS_S3_LOCKING_PROVIDER") {
+        Ok(p) if p.to_lowercase() == "dynamodb" => {
+            cfg_if::cfg_if! {
+                if #[cfg(feature = "dynamodb")] {
+                    let client = dynamodb_lock::DynamoDbLockClient::new(
+                        rusoto_dynamodb::DynamoDbClient::new(_region),
+                        dynamodb_lock::Options::default()
+                    );
+                    Ok(Some(Box::new(client)))
+                } else {
+                    Err(StorageError::Generic("dynamodb feature is not enabled".to_string()))
+                }
+            }
+        }
+        _ => Ok(None),
+    }
+}
+
+const DEFAULT_MAX_RETRY_ACQUIRE_LOCK_ATTEMPTS: u32 = 10_000;
+
+async fn rename_with_lock(
+    s3_backend: &S3StorageBackend,
+    src: &str,
+    dst: &str,
+) -> Result<(), StorageError> {
+    let lock_client = match s3_backend.lock_client {
+        Some(ref lock_client) => lock_client,
+        None => {
+            return Err(StorageError::S3Generic(
+                "dynamodb locking is not enabled".to_string(),
+            ))
+        }
+    };
```

Comment:
@xianwill @houqp So what I've came up with that is throwing an error when users call rename (used only in commit loop currently) but lock_client is none, which is either dynamodb not configured or its cargo feature turned off